src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1433 @@
+                }
+
+                return syntax;
+            }
+
+            // Given a normalized pattern (so there are only `and` and `or` patterns at the root of the tree)
+            // we traverse the binary patterns building a set of cases and reporting reachability issues
+            // on that set of cases when applicable.
+            static void analyze(BoundPattern pattern, ReachabilityAnalysisContext context)
+            {
+                if (pattern is BoundBinaryPattern binaryPattern)
+                {
+                    var currentCases = ArrayBuilder<BoundPattern>.GetInstance();
+                    analyzeBinary(currentCases, binaryPattern, wrapIntoParentAndPattern: null, context);
+                    currentCases.Free();
+                    return;
+                }
+
+                return;
+            }
+
+            static void analyzePattern(ArrayBuilder<BoundPattern> currentCases, BoundPattern pattern, Func<BoundPattern, BoundPattern>? wrapIntoParentAndPattern, ReachabilityAnalysisContext context)
+            {
+                if (pattern is BoundBinaryPattern nestedBinary)
+                {
+                    analyzeBinary(currentCases, nestedBinary, wrapIntoParentAndPattern, context);
+                }
+            }
+
+            static void analyzeBinary(ArrayBuilder<BoundPattern> currentCases, BoundBinaryPattern binaryPattern, Func<BoundPattern, BoundPattern>? wrapIntoParentAndPattern, ReachabilityAnalysisContext context)
+            {
+                if (binaryPattern.Disjunction)
+                {
+                    var patterns = ArrayBuilder<BoundPattern>.GetInstance();
+                    addPatternsFromOrTree(binaryPattern, patterns);
+
+                    // For `A1 or ... or An`, we check reachability on: `case A1`, ..., `case An` (with each wrapped as indicated by caller)
+                    int savedStackCount = currentCases.Count;
+                    foreach (var pattern in patterns)
+                    {
+                        var wrappedPattern = wrapIntoParentAndPattern?.Invoke(pattern) ?? pattern;
+                        currentCases.Add(wrappedPattern);
+                    }
+                    checkReachability(currentCases, context);
+                    currentCases.RemoveRange(index: savedStackCount, length: currentCases.Count - savedStackCount);
+
+                    // In `A1 or ... or Ai or B or ...`, we analyze `B` with `case A1:`, ..., `case Ai:` added to current cases.
+                    // That way, if `B` can be seen as multiple cases: `case B1`, ..., `case Bn`
+                    // we'll be able to check reachability on: `case A1:`, ... `case Ai:`, `case B1:`, ... `case Bn:`
+                    for (int i = 0; i < patterns.Count; i++)
+                    {
+                        analyzePattern(currentCases, patterns[i], wrapIntoParentAndPattern, context);
+                        currentCases.Add(patterns[i]);
```

Comment:
>currentCases.Add(patterns[i]); [](http://example.com/codeflow?start=24&length=30)

We used to be able to indicate that this pattern doesn't need to be reported because it was already checked. It looks like we no longer do this. Could this be a source of a duplicate warning?